File: gravity_assist_specfic_impulse.py
# import numpy as np

# Constants
G0 = 9.81  # Standard gravity, m/s^2

def calculate_thrust(specific_impulse, mass_flow_rate):
    """Calculate thrust given specific impulse and mass_flow_rate.
    Parameters:
    - specific_impulse: Specific impulse of the thruster (s).
    - mass_flow_rate: Mass flow rate of the propellant (kg/s).
    Returns:
    - Thrust produced by the thruster (N).
    """
    return specific_impulse * mass_flow_rate * G0

def estimate_power_consumption(specific_impulse, mass_flow_rate, efficiency=0.7):
    """Estimate power consumption of the thruster based on its specific impulse and mass flow rate.
    Parameters:
    - specific_impulse: Specific impulse of the thruster (s).
    - mass_flow_rate: Mass flow rate of the propellant (kg/s).
    - efficiency: Efficiency of the thruster (dimensionless, default=0.7).
    Returns:
    - Estimated power consumption (W).
    """
    energy_per_kg = specific_impulse * G0 / efficiency
    return mass_flow_rate * energy_per_kg

def optimize_thruster_parameters(gravity_assist_multiplier=1.1):
    """Find optimal thruster parameters to maximize thrust-to-power ratio.
    Parameters:
    - gravity_assist_multiplier: Multiplier to simulate gravity assist effect (dimensionless, default=1.1).
    Returns:
    - Dictionary containing optimal values for thrust, specific impulse, mass flow rate, and power consumption.
    """
    specific_impulse_range = np.linspace(2000, 5000, 50)
    mass_flow_rate_range = np.linspace(0.00005, 0.0005, 50)

    spi_grid, mass_grid = np.meshgrid(specific_impulse_range, mass_flow_rate_range)
    spi_values = spi_grid.ravel()
    mass_values = mass_grid.ravel()

    # Vectorized calculation of thrust and power consumption
    thrust_values = calculate_thrust(spi_values, mass_values) * gravity_assist_multiplier
    power_values = estimate_power_consumption(spi_values, mass_values)

    # Calculate thrust-to-power ratio and find the optimal index
    ratio_values = thrust_values / power_values
    optimal_idx = np.argmax(ratio_values)

    # Extract optimal parameters
    optimal_spi = spi_values[optimal_idx]
    optimal_mass = mass_values[optimal_idx]
    optimal_thrust = thrust_values[optimal_idx]
    optimal_power = power_values[optimal_idx]

    return {
        "Optimal Thrust": optimal_thrust,
        "Optimal Specific Impulse": optimal_spi,
        "Optimal Mass Flow Rate": optimal_mass,
        "Estimated Power Consumption": optimal_power
    }

# Running the optimization
optimal_results = optimize_thruster_parameters()
print(f"Optimal Thrust with Gravity Assist: {optimal_results['Optimal Thrust']} N")
print(f"Optimal Specific Impulse: {optimal_results['Optimal Specific Impulse']} seconds")
print(f"Optimal Mass Flow Rate: {optimal_results['Optimal Mass Flow Rate']} kg/s")
print(f"Estimated Power Consumption: {optimal_results['Estimated Power Consumption']} Watts")

File: gravity_assist_specfic_impulse.m
% Function to calculate thrust
function thrust = calculate_thrust(specific_impulse, mass_flow_rate, g0)
   % Calculate thrust given specific impulse, mass flow rate, and gravity.
   thrust = specific_impulse * mass_flow_rate * g0;
end

function power = estimate_power_consumption(specific_impulse, mass_flow_rate, efficiency, g0)
   % Estimate power consumption based on specific impulse, mass flow rate, efficiency, and gravity.
   energy_per_kg = specific_impulse * g0 / efficiency;
   power = mass_flow_rate * energy_per_kg;
end

function results = optimize_thruster_parameters(gravity_assist_multiplier, efficiency, g0)
   % Set up parameter ranges
   specific_impulse_range = 2000:500:4000;
   mass_flow_rate_range = 0.00005:0.00005:0.0005;
   
   % Vectorized calculation of thrust and power consumption
   [spi_grid, mass_grid] = meshgrid(specific_impulse_range, mass_flow_rate_range);
   thrust_values = calculate_thrust(spi_grid, mass_grid, g0) * gravity_assist_multiplier;
   power_values = estimate_power_consumption(spi_grid, mass_grid, efficiency, g0);
   
   % Calculate thrust-to-power ratio
   ratio_values = thrust_values ./ power_values;
   [~, optimal_idx] = max(ratio_values);
   
   % Extract optimal parameters
   optimal_spi = spi_grid(optimal_idx);
   optimal_mass = mass_grid(optimal_idx);
   optimal_thrust = thrust_values(optimal_idx);
   optimal_power = power_values(optimal_idx);
   
   % Store results in a container
   results = containers.Map('KeyType', 'string', 'ValueType', 'double');
   results('Optimal Thrust') = optimal_thrust;
   results('Optimal Specific Impulse') = optimal_spi;
   results('Optimal Mass Flow Rate') = optimal_mass;
   results('Estimated Power Consumption') = optimal_power;
end

% Running the optimization with gravity assist
gravity_multiplier = input('Enter gravity assist multiplier (usually > 1): ');
efficiency = 0.9; % Assuming 90% efficiency
g0 = 9.81; % Standard gravity in m/s^2
results = optimize_thruster_parameters(gravity_multiplier, efficiency, g0);

% Displaying the results
fprintf('Optimal Thrust with Gravity Assist: %.2f N\n', results('Optimal Thrust'));
fprintf('Optimal Specific Impulse: %.2f s\n', results('Optimal Specific Impulse'));
fprintf('Optimal Mass Flow Rate: %.5f kg/s\n', results('Optimal Mass Flow Rate'));
fprintf('Estimated Power Consumption: %.2f Watts\n', results('Estimated Power Consumption'));

% Optional Plot
figure
scatter(results('Optimal Specific Impulse'), results('Optimal Thrust'), 'r', 'filled');
hold on
for spi = 2000:500:4000
   thrust_curve = cellfun(@(mass) calculate_thrust(spi, mass, g0), num2cell(0.00005:0.00005:0.0005));
   plot(0.00005:0.00005:0.0005, thrust_curve, 'b');
end
xlabel('Mass Flow Rate (kg/s)');
ylabel('Thrust (N)');
title('Thrust vs Mass Flow Rate for Different Specific Impulses');
grid on;
hold off;
